semantics/language/execution/io-buffered.k

```diff
@@ -102,217 +88,190 @@ module C-IO-BUFFERED
           [structural]
      rule <k> (.K => UNSPEC("EIO6", "Attempting to access an object with thread or auto storage duration in a thread other than the one in which the object is associated.", "6.2.4:4-5"))
-               ~> readBytes(Loc:SymLoc, _, _)
+               ~> readBytes(Loc::SymLoc, _, _)
           ...</k>
-          <thread-id> ThreadId:Int </thread-id>
+          <thread-id> ThreadId::Int </thread-id>
           requires (isThreadDuration(Loc) orBool isAutoDuration(Loc))
                andBool (ThreadId =/=K getThreadId(Loc))
           [structural]
 
      // loc, size in bytes, aux list
      syntax KItem ::= "readBytes-aux" "(" SymLoc "," Int "," List "," Type ")"
-     rule readBytes-aux(Loc:SymLoc, Size:Int, Aux:List, T:Type)
-          => readByte(Loc, T)
-          ~> readBytes-aux(Loc +bytes 1, Size -Int 1, Aux, T)
-          requires Size >Int 0
-          [structural]
-     rule readBytes-aux(_, 0, Aux:List, _) => dataList(bvalues(Aux))
-          [structural]
-     rule (tv(V:CValue, T:Type) => .K)
-          ~> readBytes-aux(_, _,
-               (Aux:List => Aux ListItem(tv(V, T))), _)
+     rule readBytes-aux(_, 0, Aux::List, _) => dataList(Aux)
           [structural]
 
-     syntax List ::= bvalues(List) [function]
-     rule bvalues(ListItem(tv(K:K, _)) L:List) => ListItem(K) bvalues(L)
-     rule bvalues(.List) => .List
-
      syntax Error ::= "errorReadOutOfBounds"
                     | "errorUnseqEffects"
                     | "errorNonVolatileLVal"
 
+     syntax K ::= assertInBounds(Int, Int)
+     syntax K ::= assertSeq(SymLoc, List)
+     syntax K ::= assertVolatile(Type, Int, Type)
+
+     rule assertInBounds(Offset::Int, Len::Int) 
+          => assert(Offset <Int Len, errorReadOutOfBounds) [macro]
+     rule assertSeq(Loc::SymLoc, Locs::List) 
+          => assert(notBool (Loc in Locs), errorUnseqEffects) [macro]
+     rule assertVolatile(ObjT::Type, Offset::Int, T::Type) 
+          => assert((notBool (Volatile in getQualsAtOffset(ObjT, Offset)))
+               orBool isVolatileType(T), errorNonVolatileLVal) [macro]
+
      syntax KItem ::= readByte(SymLoc, Type)
-     rule <k> readByte(loc(Base:SymBase, Offset:Int), T:Type)
+     //TODO(dwightguth): make these rules look a little better once we have partial compilation
+     rule <k> readBytes-aux(loc(Base::SymBase, Offset::Int), Size::Int, Aux::List, T::Type)
                // TODO(chathhorn): do we really need to check this stuff every
                // byte?
-               => assert(notBool (loc(Base, Offset) in Locs),
-                    errorUnseqEffects)
-               ~> assert(Offset <Int Len, errorReadOutOfBounds)
-               ~> assert((notBool (Volatile in getQualsAtOffset(ObjT, Offset)))
-                    orBool isVolatileType(T),
-                    errorNonVolatileLVal)
-               ~> tv(V, t(.Set, no-type))
+               => assertSeq(loc(Base, Offset), Locs)
```

Comment:
@andreistefanescu is this what you had in mind? Or do you still have some issue even with this version?